server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java

```diff
@@ -53,16 +93,67 @@ public long nowInMillis() {
     }
 
     /**
-     * Returns an instance of {@link SearchExecutionContext} if available of null otherwise
+     * Returns an instance of {@link SearchExecutionContext} if available or null otherwise
      */
     public SearchExecutionContext convertToSearchExecutionContext() {
         return null;
     }
 
+    /**
+     * Returns an instance of {@link CoordinatorRewriteContext} if available or null otherwise
+     */
     public CoordinatorRewriteContext convertToCoordinatorRewriteContext() {
         return null;
     }
 
+    /**
+     * Returns the {@link MappedFieldType} for the provided field name.
+     * If the field is not mapped, the behaviour depends on the index.query.parse.allow_unmapped_fields setting, which defaults to true.
+     * In case unmapped fields are allowed, null is returned when the field is not mapped.
+     * In case unmapped fields are not allowed, either an exception is thrown or the field is automatically mapped as a text field.
+     * @throws QueryShardException if unmapped fields are not allowed and automatically mapping unmapped fields as text is disabled.
+     * @see QueryRewriteContext#setAllowUnmappedFields(boolean)
+     * @see QueryRewriteContext#setMapUnmappedFieldAsString(boolean)
+     */
+    public MappedFieldType getFieldType(String name) {
+        return failIfFieldMappingNotFound(name, fieldType(name));
+    }
+
+    protected MappedFieldType fieldType(String name) {
+        // If the field is not allowed, behave as if it is not mapped
+        if (allowedFields != null && false == allowedFields.test(name)) {
+            return null;
+        }
+        MappedFieldType fieldType = runtimeMappings.get(name);
+        return fieldType == null ? mappingLookup.getFieldType(name) : fieldType;
+    }
+
+    public IndexAnalyzers getIndexAnalyzers() {
+        if (mapperService == null) {
+            return IndexAnalyzers.of(Collections.emptyMap());
+        }
+        return mapperService.getIndexAnalyzers();
+    }
+
+    MappedFieldType failIfFieldMappingNotFound(String name, MappedFieldType fieldMapping) {
+        if (fieldMapping != null || allowUnmappedFields) {
+            return fieldMapping;
+        } else if (mapUnmappedFieldAsString) {
+            TextFieldMapper.Builder builder = new TextFieldMapper.Builder(name, getIndexAnalyzers());
+            return builder.build(MapperBuilderContext.root(false)).fieldType();
+        } else {
+            throw new QueryShardException(this, "No field mapping can be found for the field with name [{}]", name);
+        }
+    }
+
+    public void setAllowUnmappedFields(boolean allowUnmappedFields) {
```

Comment:
I wonder if we can remove these two by having a special PercolatorRewriteContext that wraps an incoming rewrite context and overrides `failIfFieldMappingNotFound()`?  For a follow-up maybe.

Reply:
Need to understand how to deal with `SearchExecutionContext#reset` calling `setAllowUnmappedFields`...maybe we can just get rid of `reset()` by having a specific `PercolatorExecutionContext`?

Reply:
To do in a followup PR.